### PR TITLE
[ci][byod/2] compute anyscale byod for release tests

### PR DESCRIPTION
**Pull request by can-anyscale**
## Why are these changes needed?
**Context:** This is a series of PR that will allow OSS release tests to run on anyscale byod images. Currently, it runs against nightly anyscale which can be unreliable. Using byod image, we will be able to run them against staging or production anyscale. Proposal doc: https://docs.google.com/document/d/12WU4vpR0jHgOwceoSgPtE4B-B-6mBonnHHiCSMmMNoY/edit#

- Move Test to its own file and class
- Logic to get the correct ray and anyscale image for a test, based on its configuration (right now just python version, in the future, cpu vs. gpu, etc.)
- A holder for @aslonnie to implement logic to build anyscale custom images given a list of ray images

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- Testing Strategy
   - [X] Unit tests
   - [X] Release tests

### CODE REVIEW

File: release/ray_release/scripts/build_pipeline.py [hunk 1]
```diff
@@ -237,5 +240,14 @@ def main(
     print(steps_str)
 
 
+def _build_anyscale_byod_images(tests: List[Tuple[Test, bool]]) -> None:
+    """
+    Builds the Anyscale BYOD images for the given tests.
+    """
+    _ = set(test.get_ray_image() for test, _ in tests)
+    # TODO(aslonnie): Build anyscale boyd images for the given ray images
```
Comment (aslonnie) [inline release/ray_release/scripts/build_pipeline.py:248]: nit: "BYOD", not "boyd"

File: release/ray_release/scripts/build_pipeline.py [hunk 2]
```diff
@@ -151,6 +152,8 @@ def main(
             "Empty test collection. The selected frequency or filter did "
             "not return any tests to run. Adjust your filters."
         )
+    logger.info("Build anyscale BYOD images")
+    _build_anyscale_byod_images(filtered_tests)
```
Comment (krfricke) [inline release/ray_release/scripts/build_pipeline.py:156]: I'm not sure if this should happen in the pipeline builder script. It presumably takes quite a bit of time to build these images. Should this happen here? Are tests sharing the same image?

**Review by krfricke — COMMENTED** [review]
Do we have a doc or something on how the end state should look like?

File: release/ray_release/scripts/build_pipeline.py (hunk 2 above)
Comment (can-anyscale) [reply release/ray_release/scripts/build_pipeline.py:156]: Sorry for the lack of details. Hopefully this give more context: https://docs.google.com/document/d/12WU4vpR0jHgOwceoSgPtE4B-B-6mBonnHHiCSMmMNoY/edit#

To your question, this doesn't rebuild ray image, only anyscale-ray image so took about 10 minute to finish. But yes, I want to build the image once and reuse them across all tests (a nice side effect is that the customer environment build that take about 10 minute for each test will be 0).

Comment (krfricke) [reply release/ray_release/scripts/build_pipeline.py:156]: OK, thanks for the context. That sounds good to me

File: release/ray_release/scripts/build_pipeline.py [hunk 3]
```diff
@@ -237,5 +240,14 @@ def main(
     print(steps_str)
 
 
+def _build_anyscale_byod_images(tests: List[Tuple[Test, bool]]) -> None:
+    """
+    Builds the Anyscale BYOD images for the given tests.
+    """
+    _ = {test.get_ray_image() for test, _ in tests}
+    # TODO(aslonnie): Build anyscale byod images for the given ray images
+    return
```
Comment (krfricke) [inline release/ray_release/scripts/build_pipeline.py:249]: Suggestions: Let's move this to a separate package e.g. `ray_release.byod` which also contains the dockerfile, dependencies, etc

**Review by krfricke — APPROVED** [review]
Thanks for the context. Looks good to me. I'd suggest we separate the logic into a submodule, but otherwise happy to proceed

File: release/ray_release/scripts/build_pipeline.py (hunk 3 above)
Comment (can-anyscale) [reply release/ray_release/scripts/build_pipeline.py:249]: I like that!

File: release/ray_release/config.py [hunk 4]
```diff
@@ -6,19 +6,14 @@
...
 from ray_release.util import DeferredEnvVar, deep_update
 
 
-class Test(dict):
-    """A class represents a test to run on buildkite"""
-
-    pass
-
-
 class TestDefinition(dict):
```
[the remark this answers is not in the document]
Comment (can-anyscale) [reply release/ray_release/config.py:17]: Sure thing!

**Review by aslonnie — APPROVED** [review]

**Pull request merged**